base/serialize.jl

```diff
@@ -213,14 +213,17 @@ function serialize{T,N,A<:Array}(s::SerializationState, a::SubArray{T,N,A})
 end
 
 function trimmedsubarray{T,N,A<:Array}(V::SubArray{T,N,A})
-    dest = Array{eltype(V)}(trimmedsize(V))
+    dest = Base.allocate_for(Array{eltype(V)}, V.parent, trimmedsize(V))
     copy!(dest, V)
     _trimmedsubarray(dest, V, (), V.indexes...)
 end
 
-trimmedsize(V) = index_lengths(V.parent, V.indexes...)
+trimmedsize(V) = index_shape(Val{false}, V.parent, V.indexes...)
```

Comment:
Hah, if I'd noticed the `A<:Array` constraint in that serialization method, I could have saved myself a lot of effort!